sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+        case LeftSemi => Array(leftKeys.leftSemiJoinGatherMap(rightKeys, false))
+        case LeftAnti => Array(leftKeys.leftAntiJoinGatherMap(rightKeys, false))
+        case FullOuter => leftKeys.fullJoinGatherMaps(rightKeys, false)
+        case _ =>
+          throw new NotImplementedError(s"Joint Type ${joinType.getClass} is not currently" +
+              s" supported")
+      }
+      makeGatherer(maps, leftData, rightData)
+    }
+  }
+
+  private def joinGatherMapLeftRight(
+      leftKeys: ColumnarBatch,
+      leftData: LazySpillableColumnarBatch,
+      rightKeys: ColumnarBatch,
+      rightData: LazySpillableColumnarBatch): Option[JoinGatherer] = {
+    withResource(GpuColumnVector.from(leftKeys)) { leftKeysTab =>
+      withResource(GpuColumnVector.from(rightKeys)) { rightKeysTab =>
+        joinGatherMapLeftRight(leftKeysTab, leftData, rightKeysTab, rightData)
+      }
+    }
+  }
+
+  private def joinGatherMap(
+      buildKeys: ColumnarBatch,
+      buildData: LazySpillableColumnarBatch,
+      streamKeys: ColumnarBatch,
+      streamData: LazySpillableColumnarBatch): Option[JoinGatherer] = {
+    buildSide match {
+      case GpuBuildLeft =>
+        joinGatherMapLeftRight(buildKeys, buildData, streamKeys, streamData)
+      case GpuBuildRight =>
+        joinGatherMapLeftRight(streamKeys, streamData, buildKeys, buildData)
+    }
+  }
+
+  private def joinGatherMap(
+      buildKeys: ColumnarBatch,
+      buildData: LazySpillableColumnarBatch,
+      streamCb: ColumnarBatch): Option[JoinGatherer] = {
+    withResource(GpuProjectExec.project(streamCb, boundStreamKeys)) { streamKeys =>
+      withResource(GpuProjectExec.project(streamCb, boundStreamData)) { streamData =>
+        joinGatherMap(buildKeys, buildData,
+          streamKeys, new LazySpillableColumnarBatch(streamData, spillCallback))
+      }
+    }
+  }
+
+  override def hasNext: Boolean = {
+    var mayContinue = true
+    while (nextCb.isEmpty && mayContinue) {
+      val startTime = System.nanoTime()
+      if (gathererStore.exists(!_.isDone)) {
+        nextCb = nextCbFromGatherer()
+      } else if (stream.hasNext) {
+        // Need to refill the gatherer
+        gathererStore.foreach(_.close())
+        gathererStore = None
+        withResource(stream.next()) { cb =>
+          streamTime += (System.nanoTime() - startTime)
+          gathererStore = joinGatherMap(builtKeys.getBatch, builtData, cb)
+        }
+        nextCb = nextCbFromGatherer()
+      } else if (initialJoin) {
+        withResource(GpuColumnVector.emptyBatch(streamAttributes.asJava)) { cb =>
+          gathererStore = joinGatherMap(builtKeys.getBatch, builtData, cb)
+        }
+        nextCb = nextCbFromGatherer()
+      } else {
+        mayContinue = false
+      }
+      totalTime += (System.nanoTime() - startTime)
+      initialJoin = false
+    }
+    if (nextCb.isEmpty) {
+      // Nothing is left to return so close ASAP.
+      close()
```

Comment:
Is there a risk of a double-close error on a cudf column vector if we close here and the caller separately closes after this returns false?

Reply:
Yes, but double frees are OK, although I will protect against it anyway to future proof it.